networking_cisco/plugins/cisco/db/l3/l3_router_appliance_gbp_db.py

```diff
@@ -0,0 +1,82 @@
+#    under the License.
+
+from networking_cisco.plugins.cisco.db.l3 import (
+    l3_router_appliance_db as l3_r_db)
+from networking_cisco.plugins.cisco.l3.drivers import driver_context
+
+from neutron.common import constants as l3_constants
+from neutron.common import exceptions as n_exc
+from neutron.i18n import _LW
+
+from oslo_config import cfg
+from oslo_log import log as logging
+
+FLOATINGIP_STATUS_ACTIVE = l3_constants.FLOATINGIP_STATUS_ACTIVE
+
+LOG = logging.getLogger(__name__)
+
+
+class L3RouterApplianceGbpDBMixin(l3_r_db.L3RouterApplianceDBMixin):
+    """Mixin class implementing Neutron's routing service using appliances.
+
+    This class overrides the base superclass, and should only be used for
+    integrations with ACI using the Group Based Policy workflow.
+    """
+
+    def create_floatingip(self, context, floatingip,
+                          initial_status=FLOATINGIP_STATUS_ACTIVE):
+        """Override of parent class's create_floatingip.
+
+        This override of the parent class's create_floatingip assumes
+        is needed for GBP workflow, as the pre-/post-commmit calls for
+        creating the floating IP must be peformed in a loop with the database
+        call in the parent class.
+        """
+
+        res = None
+        fip = floatingip['floatingip']
+        if not fip.get('subnet_id'):
+            # NOTE: default router type must be ASR1k
+            router_type_name = cfg.CONF.routing.default_router_type
+            driver = self._get_router_type_driver(context,
+                                                  router_type_name)
+            if driver:
+                fip_ctxt = driver_context.FloatingipContext(floatingip)
+                driver.create_floatingip_precommit(context, fip_ctxt)
+                nat_pool_list = getattr(context, 'nat_pool_list', [])
+                for nat_pool in nat_pool_list:
+                    if not nat_pool:
+                        continue
+                    fip['subnet_id'] = nat_pool['subnet_id']
+                    try:
+                        res = super(l3_r_db.L3RouterApplianceDBMixin,
+                                    self).create_floatingip(context, floatingip)
+                        router_ids = ([res['router_id']]
+                                      if res['router_id'] else [])
+                    except n_exc.IpAddressGenerationFailure as ex:
+                        LOG.warning(_LW("Floating allocation failed: %s"),
+                                    ex.message)
+                    if res:
+                        break
+        if not res:
+            info = super(l3_r_db.L3RouterApplianceDBMixin,
+                         self).create_floatingip(context,
+                                                 floatingip, initial_status)
+            router_ids = [info['router_id']] if info['router_id'] else []
+        context.res = res
+        if driver:
```

Comment:
Note that this skips the immediate parent and calls the grandparent method, so the precommit won't be called. The case where subnet_id is true is there for when the GBP code invokes create_floatingip, pre-populating the call with the subnet_id attribute (i.e. different use-case/caler). Don't worry if this all is confusing -- it took me a while to understand this (there's also a monkey-patch for the grandparent class in gbpservice/neutron/extensions/patch.py).